docs/how-to/contribute-tests.rst

```diff
@@ -0,0 +1,118 @@
+Implementing Tests in Batavia
+=============================
+
+Basic Test Structure
+--------------------
+
+Batavia's job is to run a browser-compatible Python compiler, which takes valid Python as input and runs it.
+Therefore, tests should test that the output of the Batavia compiler matches the output of CPython::
+
+    print('Hello')  # Code to test
+    Hello               # CPython output
+    Hello               # Batavia output
+    # Outputs match. Test passes!
+
+This test structure is simple and effective. It's used in almost every test we've written.
+
+Adding Tests
+------------
+
+In many cases, existing tests will not cover everything. Feel free to add your own!
+
+The tests corresponding to Batavia implementations of built-ins are available inside
+``tests/builtins``. The Batavia test infrastructure includes a system to check the compatibility of
+JavaScript implementation of Python with the reference CPython implementation.
+
+These tests all derive from ``TranspileTestCase``, which handles running your code in both interpreters
+and comparing outputs. For an example, look at the ``test_bool.py`` file in ``tests/builtins``. You 
+will see two classes with test cases, ``BoolTests`` and ``BuiltinBoolFunctionTests``. Both derive 
+from ``TranspileTestCase``.
+
+Let's look at some test code that checks if a the Batavia implementation of ``bool`` can handle a
+bool-like class that implements ``__bool__``.
+
+.. code-block:: Python
+
+    def test_bool_like(self):
+        self.assertCodeExecution("""
+            class BoolLike:
+                def __init__(self, val):
+                    self.val = val
+
+                def __bool__(self):
+                    return self.val == 1
+            print(bool(BoolLike(0)))
+            print(bool(BoolLike(1)))
+            """)
+
+The ``assertCodeExecution`` method will run the code provided to it in both implementations. This
+code needs to generate some output so that the output can be compared, hence the need to print the
+values. **Code that is not being printed is not being tested.**
```

Comment:
Do we want to mention some of the edge cases, like code that produced equivalent output in a different order (e.g., sets and dict keys) will also fail, so tests should be structured to compensate - either by restructuring the test (e.g., convert set to a sorted list and compare the type of the original value, rather than using set comparison), or using test output subsitutions?

Reply:
Added a bit for this.